[v1.7] routing: Fix route collisions in AWS ENI

christarazi opened the pull request:
# Problem

To recap, the [ENI IPAM](https://docs.cilium.io/en/v1.7/concepts/ipam/eni/#ipam-eni) mode allocates a per-ENI (interface) routing table. Each endpoint in this IPAM mode has two ip rules, "ingress" and "egress", and 2 routes. These 4 items will be created under same table ID. This table ID is the ifindex of the interface. It is possible that the ifindex of the interface is equal to 253, 254, or 255 [1], which are reserved for the main routing table. Therefore, it is possible for us to inject routes (and rules) reserved for the main routing of the system, which can obviously cause connectivity disruption. 

For more on this issue, see https://github.com/cilium/cilium/issues/14336.

# Proposed Solution

The solution is to compute the proper table ID by taking the "interface number" and adding an offset of 10, to avoid the possibility of colliding with the main routing table.

 The "interface number" is given to us by the AWS API for the interface (`CiliumNode.Status.ENI.ENIs[*].Number`). This number is sequential meaning the first ENI has "interface number" of 1, and so on. The max number of ENIs on a given node is 15. So, the max possible table ID is 25. Therefore, we avoid the possibility of colliding with the main routing table. An offset of 10 was chosen because it avoids colliding with other system components which may allocate lower numbered table IDs and because it will be smaller than the main routing table ID range. 

In order to rollout this solution, we must migrate the old ENI datapath on upgrade. Meaning, we must fixup all the old rules and routes which use the old table ID scheme, to the new table ID scheme. This includes creating new rules, routes, then deleting the old rules and routes only if all previous steps succeeded. This is a sensitive operation, as any failure could render the node's connectivity useless. If we encounter any errors during the migration process, we will attempt to rollback to the previous table ID scheme, to hopefully restore the node connectivity. 

If the user wants to downgrade Cilium, the user must first set `--egress-multi-home-ip-rule-compat=true` and restart the agent. Cilium will see that the ENI datapath has been migrated and with the flag `true`, it will revert the migration. After this is complete, the user may downgrade their version of Cilium. The docs have been updated to mention this.

Note, while this does affect Azure IPAM mode as well, v1.7 does not support Azure mode, therefore support for it is missing in this PR. In the forward-ported version of this PR, it will be included.

[1]: See `ip-route(8)`
___

This is a direct backport to the v1.7 branch.

See commit msgs.

Related: https://github.com/cilium/cilium/issues/14336

christarazi committed:
routing: Fix route collisions in AWS ENI

This commit fixes a potential route collision in AWS ENI IPAM modes,
where the ifindex could equal the main routing table ID (from 253-255)
[1], causing traffic to be subject to these routes incorrectly. This is
admittedly rare, but we've seen this from a user report. The impact is
that most traffic on the node is suddenly blackholed.

To fix this, we say that each device or interface (ENI) will have their
own dedicated routing table. The table ID will start with an offset of
10 because it is highly unlikely to collide with the main routing table
ID (from 253-255). We grab the number associated with the ENI device
(`Number`) and add the offset. For example, if we have an ENI device
"eni-0" which has a `Number` of 5, then the table ID will be 10 + 5 =
15.

Another important piece to note is that only the egress rule will reside
inside the per-device tables, whereas the ingress rule will stay in the
main routing table. This is because we want the main routing table to
hold the routes to the endpoint.

Moving forward, the ENI datapath will now create rules under a new
egress priority value (RulePriorityEgressv2), as long as the
egress-multi-home-ip-rule-compat flag is false. If it's true, then the
datapath will create rules under the original egress priority value
(RulePriorityEgress). This helps disambiguate when running with the
older or newer ENI datapath.

See https://github.com/cilium/cilium/issues/14336.

[1]: See ip-route(8)

Reported-by: Vlad Ungureanu <vladu@palantir.com>
Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon, option: Add flag egress-multi-home-ip-rule-compat

This flag is needed to control the behavior of Cilium when it starts up
under ENI mode.

If the flag is false, meaning "do not maintain compatibility", then
Cilium will attempt to migrate the ENI datapath (`ip rule`s and routes)
created under the aforementioned IPAM mode to a new table ID scheme. The
table ID refers to the Linux routing policy database tables, aka
"routing table".

If the flag is true, meaning "maintain compatibility", then Cilium will
not attempt to migrate the ENI database under the aforementioned IPAM
mode to the new table ID scheme. It will continue to use the original
scheme. Additionally, when the flag is true and Cilium finds the rules
under the newer scheme (by checking the priority of the rule), it will
attempt to migrate back to the original scheme. This allows downgrading
Cilium without affecting connectivity.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
routing: Add ENI route table migration logic

This commit will fixup the ENI datapath depending on the
egress-multi-home-ip-rule-compat flag (see previous commits for
context). The migration logic supports both upgrading and downgrading
the ENI datapath.

This logic must run on startup before the API is served and before the
health endpoint is created, so that no endpoints are prematurely
crreated before Cilium has had the chance to migrate the entire
datapath.

See https://github.com/cilium/cilium/issues/14336.

Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
api: Extend IPAM to accept interface number

This is needed in ENI mode. In upcoming commits, the interface number
(ENI.Number) will be used to compute the per-ENI table ID in order to
store rules and routes for the ENI datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
cni, routing: Plumb interface number

In the previous commit, we added the interface number to the IPAM
response for ENI mode. This commit plumbs this new field into the CNI to
set up the ENI datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -1394,6 +1394,10 @@ definitions:
           The UUID for the expiration timer. Set when expiration has been
           enabled while allocating.
+      interface-number:
+        type: string
```
Given the way we intend to use this, should it be a `type: integer` instead? Or did you find that it's too awkward to deal with integer types from the API?

christarazi replied:
Yeah, I went with string throughout because it was too awkward to use anything else. Specifically for the openapi.yaml, for some reason, I wasn't even allowed to use an integer here and `make generate-api` failed with "you must use `string` or `array`".

joestringer replied:
Swagger, you had one job... 

SGTM. Not ideal but I had a suspicion that this might be the answer. The only reason I thought there might be another way was that I saw some use of `type: integer` elsewhere in the openapi.yaml.

brb replied:
Maybe it's because the other property fields are `string` or `array` :-/

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -58,7 +59,8 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, masq bool) error {
 
-	// Route all traffic to the ENI address via the main routing table
+	// Route all traffic to the ENI address via the main routing table.
+	// Ingress rules are an exception to the above comment.
```
I found this comment confusing, isn't all traffic to the ENI address considered "ingress"? Maybe you meant egress? But then it's not destined to the ENI address.

christarazi replied:
Ah, I recall now what happened. I initially had a different implementation than what this PR has and when I cherry-picked the commit here, I forgot to update this comment and so it doesn't refer to anything correctly. You're correct that if I change it to say "egress" it would be accurate.

tgraf reviewed (approved):
LGTM

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -118,12 +122,13 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, masq bool) error {
+// know the table ID at delete-time. However, the ingress rule is always in the
+// main routing table. The egress rule is a best-effort search, but it is
+// created under a specific priority, which is unlikely to have collisions with
+// other rules on the system, or so we hope.
```
The *rule* in the routing rules list points to a table, where the *routes* live right? I think the existing comment is a little confusing given "routes within these rules" but my understanding is that logically there's a linked list of rules (with priority) which then point into some kind of map/array of tables that decide how to forward traffic.

The ingress/egress distinction as far as I can tell is purely a Cilium construct where we dedicate ingress priority in this initial linked list as 20 and egress priority as 110 , then there's a few other stock standard rules before/in-between/after, including the final priority 32766-32767 rules at the end which point to the main/default routing tables.

I tried to reword a couple of times but got a bit confused about the main point here, I think we're trying to say something like:

> Note that one or more IPs may share the same route table, as identified by the interface number of the corresponding device. This function only removes the egress rules to disconnect the per-ENI egress routes from a specific local IP, and does not remove the corresponding route table as other IPs may still be using that table.

And maybe,

> The search for the egress rule corresponding to this IP is a best-effort based on the egress priority that Cilium uses, which we assume full control over. That is, if any other component in the system uses the same priority, this function would delete all such rules as well.

(Well I don't know whether the "delete all such rules" part is correct, but having a known consequence or even worst case just "behaviour is undefined if this happens" might be nice detail to point out here.)

christarazi replied:
> The rule in the routing rules list points to a table, where the routes live right? I think the existing comment is a little confusing given "routes within these rules" but my understanding is that logically there's a linked list of rules (with priority) which then point into some kind of map/array of tables that decide how to forward traffic.

Yes you're correct and that was what I was intending to express by saying "within" to denote some sort of linkage between the two. Your wording is much clearer and more concise so I'll borrow it :). Modifying it to mention ingress rules as well.

> (Well I don't know whether the "delete all such rules" part is correct, but having a known consequence or even worst case just "behaviour is undefined if this happens" might be nice detail to point out here.)

The deletion matches on the IP and the priority. If we find more than one rule matching, we skip deletion altogether. So it's not accurate to say that we will delete all rules with a certain priority.

joestringer reviewed (changes requested):
Some minor nits that I think would make the code more readable below.

The solution itself LGTM.

Two other remaining avenues for followup:
* I'd guess Azure could be affected by the same issue?
* Upgrade/downgrade. We had discussed some solutions out-of-band but my understanding is that this current implementation is sufficient assuming that you cordon the nodes prior to upgrade.

On the latter point, I'm not entirely sure how cleanup of previous rules/tables would happen during upgrade. Even if you cordon the node and vacate all workloads, will the old route tables / rules be removed? What about the ones for the cilium-health endpoint?

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -58,7 +59,8 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, masq bool) error {
 
-	// Route all traffic to the ENI address via the main routing table
+	// Route all traffic to the ENI address via the main routing table. Egress
+	// rules are an exception.
```
I still don't get "all traffic *to* the ENI address and then "Egress rules are an exception", the egress rules don't have the ENI address as a destination right?

christarazi replied:
Reword. I think what it was intending to say is that the traffic is routed to the EP IP rather than the "ENI address". The "egress rules are an exception" part is referring to the ingress rules associated with the main routing table, which is not the case for egress rules, hence the comment. I also clarified this part.

christarazi committed:
linux_defaults: Add RulePriorityEgressv2

This new priority value is vital for disambiguating which rules are
still under the old scheme. Without this, upgrading to the new scheme
would be difficult, as we aren't able to identify which rules have been
fixed up [1]. Furthermore, this would also allow us to enable downgrades
from the new scheme, because we would be able to identify which rules
need to be modified.

[1]: https://github.com/cilium/cilium/issues/14336

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1338,6 +1339,14 @@ func runDaemon() {
 
 	metricsErrs := initMetrics()
 
+	if option.Config.IPAM == option.IPAMENI {
```
This is occurring as basically the last thing before serving the API, were there any particular considerations you had for putting it here rather than any other particular earlier point?

I was wondering about whether it should be enforced to occur after endpoint restore, but even that I don't think actually matters (since configuration of these bits typically occurs in the CNI plugin).

christarazi replied:
I put it here because it was the easiest thing to do cognitively -- "I know that this logic must run before the API is served". And so I put it in the most obvious spot. :sweat_smile:  

If it makes sense elsewhere, happy to move it. I realize if everyone had the same logic as me, then the ordering of operations of the daemon init might get out of whack.

joestringer replied:
Something's gotta be last :-)

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,242 @@
+
+type ciliumNodeGetter interface {
+	Get(name string, opts v1.GetOptions) (*v2.CiliumNode, error)
+}
```
Organizationally it seems a little odd to include k8s and CRD APIs from the `datapath/linux/routing package`. At first glance I wonder whether it'd be cleaner for resolving dependencies to expose a lower-level API instead, eg something that allows `GetInterfaceNumberByMac()`:

```suggestion
type InterfaceDB interface {
	GetInterfaceNumberByMAC(mac string) (int, error)
}
```

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,242 @@
+		return nil
+	}
+
+	if err := replaceDefaultRoutes(old, new); err != nil {
```
`replace` here suggests that the routes in the `from` table will disappear and equivalent ones will appear in the `to` table. But if I read correctly, what is actually happening is we iterate the `from` table and then perform what is effectively an "Upsert" into the new table, potentially replacing existing routes in the destination table (that's what the "Replace" part in `RouteReplace()` is about, `unix.NLM_F_REPLACE` is set so the operation will replace any conflicting routes).

I also note the use of the words "Default routes" which has a well-known meaning in networking, specifically regarding the route for `0.0.0.0/0` (which matches all traffic). But from the function it appears to be copying all routes regardless of the prefix.

Super minor nit but if I'm following the above correctly then for readability I wonder if it would be clearer to reflect that in the naming of this function somehow, ie

```suggestion
	if err := copyRoutes(old, new); err != nil {
```

christarazi replied:
That's a good point. I like your suggestion better.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,242 @@
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress); err != nil {
```
When updating the routes or creating new rules above, we only replace the ones with the `old` table number. But now for deleting the old rules (+routes below), this is removing all rules/routes at the priority regardless of table number. This seems a bit inconsistent and could theoretically undo some of the careful ordering guarantees we're trying to make here for hitless upgrade.

christarazi replied:
Good call, will update this.

joestringer reviewed (changes requested):
I'll look over the test as well now but I already have a few preliminary comments on the new migration code below.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,275 @@
+func MigrateENIRoutingTable(getter ciliumNodeGetter) []error {
+	if option.Config.IPAM != option.IPAMENI {
+		return nil
+	}
```
We already have this check in the daemon, do we need it here? Might help to eliminate a couple of setup/teardown bits in the testsuite.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,275 @@
+		return errors
+	}
+
+	for _, r := range rules {
```
I wonder, is it possible to dump all rules in priority `RulePriorityEgress` and then iterate those specifically, rather than dumping all possible rules and iterating all of them to find the ones we care about?

christarazi replied:
Yes it is possible and I've [implemented support](https://github.com/vishvananda/netlink/pull/538) for that in netlink, but in v1.7, our netlink version does not include it. I believe we have this capability in v1.8 and onwards.

joestringer reviewed (commented):
Looks good. I wondered whether there's some more filtering we could do to reduce the number of rules being dumped out of the kernel (eg by requesting with a more specific filter of matching rules, ie "give me only the rules with priority X"), but I didn't look so closely at the libraries/kernel API to figure out whether that's possible.

christarazi commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. If table ID is comprised of the number associated with an
```
Typo, fixup after CI runs

christarazi commented on pkg/datapath/linux/routing/migrate_test.go:
```diff
@@ -0,0 +1,148 @@
+		"to", "all",
+		"table", fmt.Sprintf("%d", index),
+		"priority", fmt.Sprintf("%d", linux_defaults.RulePriorityEgress)).Run()
+	c.Assert(err, IsNil)
```
Not sure what to do with this

joestringer replied:
I'm genuinely interested in the underlying issue here, but as far as the immediate problem of writing test validation code, I believe `ip` is typically available in all environments where we run the privileged tests so I'm fine with just sticking with the uncommented implementation and we can dig separately on it.

FWIW I also tried this locally and found the same behaviour. There's likely some difference between how the library prepares the kernel netlink request vs. what `ip rule` does, and it may even be a netlink library bug or something. But from a pragmatic perspective, I don't think there's any immediate issue with just shelling out here.

joestringer replied:
^ We could file a bug to investigate further and reference it in the code like `// GH-XXXX: Must shell out here due to netlink request resulting in EINVAL.`

christarazi replied:
Done.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+			// ifindex (LinkIndex).
+			newTable, err := retrieveTableIDFromIfIndex(routes[0].LinkIndex, getter)
+			if err != nil {
+				err = fmt.Errorf("failed to retrieve new table ID: %w", err)
```
Please encode `routes[0].LinkIndex` in the error

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+func retrieveTableIDFromIfIndex(ifindex int, getter interfaceDB) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, err
```
Please encode a message here so we understand why `retrieveTableIDFromIfIndex()` failed. Otherwise we will see:
```
failed to retrieve new table ID: <error>
```

while <error> is outside of our control so we won't know what failed exactly.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(link.Attrs().HardwareAddr.String())
+	if err != nil {
+		return -1, err
```
Same here, please include something like `"failed to find ENI with MAC <MAC> in CiliumNode resource"`

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+
+	for _, r := range routes {
+		route := r
+		route.Table = to
```
You probably want to skip `r.Scope == netlink.SCOPE_LINK` to not insert the routes already inserted a second time.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+		}
+	}
+
+	for _, r := range candidates {
```
Not sure why this 2nd loop is needed and why this is not just embedded into the above loop.

christarazi replied:
I did this to guarantee that the link scoped route is inserted first. I didn't want to complicate the logic by putting it all into one loop. I figured the performance difference is probably negligible. Let me know if you think otherwise.

joestringer replied:
^ Worth noting this is specifically to address the problem where we attempt to inject a route with next-hop of the gateway first, which will be rejected by the kernel as `-ENETUNREACH` because it doesn't know how to direct traffic hitting the route towards the gateway.

IE we need to add the link-local scope route to the gateway first, then routes that depend on that as a next-hop later.

EDIT: Maybe adding this last sentence as a comment would make this more obvious in the code.

joestringer replied:
Looking at this code again in context, this is the rules function not the routes function so I'm not sure I understand @christarazi 's response here. We don't need to create a list of candidates and then iterate them, we can just add rules in-place while iterating the list in the first place.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress, old); err != nil {
+		return err
+	}
+	if err := deleteOldRoutes(old); err != nil {
```
If we succeeded with the deletion of the rules then we can treat this as a warning and not fatal. It's still somewhat serious as there can be conflict in the future but we should treat it less serious than errors above. We don't have to fully revert here I think.

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,19 @@ func runDaemon() {
+	// being created.
+	if option.Config.IPAM == option.IPAMENI {
+		if errs := linuxrouting.MigrateENIRoutingTable(&interfaceNumberGetter{}); len(errs) > 0 {
+			log.WithField("errors", errs).
```
Maybe we should rethink this a bit to provide better guarantees:
1. I think we want to fatal if we can't do successful migration. Datapath flow will continue and user can roll back immediately.
2. Once we have done migration but have not enabled it yet, we want to have a single atomic operation such as removing the old rules that is incredibly likely to succeed. This operation is likely reversible so we can re-add the rules and then still fatal to indicate that migration is not possible so user can rollback.

I'll add comments separately to indicate what it would take to get there.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+		return errors
+	}
+
+	for _, r := range rules {
```
Let's split this entire loop into two phases:

Phase 1: Set up the new tables and rules. This short be done in a first iteration of all rules. If this fails , we should cleanup and Fatal. THe system should look the same as before.

Phase 2: Delete the old rules to swap over to the new model. If this fails, we should re-add the deleted rules and then Fatal as well.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
```
You already have the list of routes for the old table, you can pass it into `migrateOnUpgrade()`

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+		return nil
+	}
+
+	if err := copyRoutes(old, new); err != nil {
```
The routes can be passed in here as well

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+func copyRoutes(from, to int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
```
This means you can avoid retrieving the routes again here. That is one less kernel interaction that can fail.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+func findRulesByPriority(prio int) ([]netlink.Rule, error) {
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
```
I would retrieve the list of rules once in the beginning and then use that list instead of doing more netlink calls. Each of them can fail and you have to handle the error.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,270 @@
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress, old); err != nil {
```
This operation should be in phase 2 and in a second iteration of all rules and then on failure, it should re-add the rules.

tgraf commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -23,6 +23,12 @@ const (
+	// interface tables. Each ENI / Azure interface will have its own table
+	// starting with this offset. It is 10 because it is unlikely to collide
+	// with the main routing table which is between 253-255. See ip-route(8).
+	RouteTableInterfacesOffset = 10
```
FYI: This is already used in previous commits.

christarazi replied:
Thanks, fixed.

brb replied:
Wouldn't this collide with the IPSec table (see above `RouteTableIPSec`)? Or is IPSec mode mutually exclusive with ENI mode?

twpayne replied:
I think it's OK as the maximum interface number is currently 15, so the maximum route table number will be 25, which is well short of the 200 required to conflict with `RouteTableIPSec`. It may well be worth updating the comment, and even adding a check in the code however.

brb replied:
OK, so my question is it guaranteed that the ENI Number will always be in the [1;15] even after multiple create and delete?

tgraf replied:
Yes

christarazi replied:
Thanks. Yes that's worth adding. Will do

tgraf reviewed (changes requested):
This is already very close. Some nits and one bigger change that we should discuss to provide better guarantees on the upgrade/migration path.

christarazi committed:
revert: Add ability to extend the revert stack

This is useful to aggregate the items to revert in one stack, so that it
can all be done at once.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
linux_defaults: Add RouteTableInterfacesOffset

This new value is the table ID for the per-ENI routing tables in the new
ENI datapath. Upcoming commits will use this value and implement the new
datapath.

See https://github.com/cilium/cilium/issues/14336.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,24 @@ func runDaemon() {
+		case err != nil && revErr == nil:
+			log.WithError(err).Error("Failed to migrate ENI datapath. The original datapath has been reinstated. Connectivity has not been affected.")
+		case err == nil && revErr == nil:
+			log.WithError(err).Error("Migration of ENI datapath successful")
```
Oops, this should not be an error.

christarazi commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,391 @@
+	return revertStack, nil
+}
+
+func onUpgradeDeleteRulesAndRoutes(
```
Add godoc

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,24 @@ func runDaemon() {
+		err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
+		switch {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. Connectivity is not guaranteed.")
```
What would we recommend the user do if they hit this? If we can suggest a mitigation here that would be ideal (close the loop for the user between seeing the issue and figuring out what they should do next when their pods are in crashloopbackoff during upgrade).

christarazi replied:
I'm not exactly sure what the mitigation is. This case is obviously the worst-case. We failed to migrate and we failed to reinstate the original datapath configuration. It could have failed to reinstate completely or could have failed on the last operation. Besides manual intervention, I'm not sure what we can recommend. Thoughts?

brb replied:
Can we include some pointers into the log message how to do the manual intervention? Or do possible errors already include this info?

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,24 @@ func runDaemon() {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. Connectivity is not guaranteed.")
+		case err != nil && revErr == nil:
+			log.WithError(err).Error("Failed to migrate ENI datapath. The original datapath has been reinstated. Connectivity has not been affected.")
```
```suggestion
			log.WithError(err).Warning("Failed to migrate ENI datapath. The original datapath has been reinstated. Connectivity has not been affected.")
```

I think we're trying to hint that something is off, but the cluster should otherwise continue most likely unaffected (other than the possibility of hitting the bug that this PR is trying to resolve). Seems like Warning is enough severity for that kind of announcement to the user.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,396 @@
+// that equals the main routing table number (253-255), causing the rules and
+// routes to be created using the wrong table ID, which could end up
+// blackholing most traffic on the node.
+func MigrateENIDatapath(getter interfaceDB) (error, error) {
```
nit: Given the two errors returned, it would be a nice godoc explanation to describe what they mean for the caller to interpret.

christarazi replied:
Yep, added in incremental diff

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1757,3 +1778,37 @@ func checkNodePortAndEphemeralPortRanges() error {
+// It retrieves the number associated with the ENI device for the given MAC
+// address. The interface number is retrieved from the CiliumNode resource, as
+// this functionality is needed for ENI mode.
+func (in *interfaceNumberGetter) GetInterfaceNumberByMAC(mac string) (int, error) {
```
I'd like to be able to move this out of daemon as well, but we probably want to wait until we figure out exactly how we implement this piece for Azure IPAM as well. No action requested today.

christarazi replied:
Agreed, it doesn't belong here, but I couldn't quite think of a better place.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,396 @@
+				return netlink.RouteDel(&route)
+			})
+
+			break
```
I think the break here is kinda negligible. If the route table matches our expectations, then breaking early doesn't make much difference to performance. We're talking maybe stop iterating at 1/2 routes. If a user ended up doing something more interesting with the routes behind the scenes that leads there to be multiple link-scoped routes then this would skip some arbitrary subset of such routes. So in the common case it doesn't make a difference, but in the more complicated case it seems like it would be potentially inconsistent. :shrug:

christarazi replied:
Good point, will remove.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,396 @@
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
+			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
+		}
```
Seems like the revertstack will be run twice in this case due to lines 152-155, which is highly likely to cause any failure in the second portion to result in error logs here.

christarazi replied:
Fixed.

brb reviewed (commented):
Just one question on ENI Number recycling. Anyway, double ACK on clean code and history!

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,23 @@ func runDaemon() {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. Connectivity is not guaranteed.")
+		case err != nil && revErr == nil:
+			log.WithError(err).Warning("Failed to migrate ENI datapath. The original datapath has been reinstated. Connectivity has not been affected.")
```
We should say something along the lines of `"Fix the error and retry"` to make sure the user understands that the error should be temporary and it should be possible to fix it somehow.

christarazi replied:
Clarified this msg and also the above msg to indicate manual intervention is required.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,397 @@
+	}
+
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
```
Is this supposed to be `revertErr`?

christarazi replied:
Good catch, yes.

tgraf commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,397 @@
+	}
+
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
```
Is this supposed to be revertErr?

christarazi replied:
Good catch, yes.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1331,26 @@ func runDaemon() {
+				"Connectivity has not been affected. Consider investigating the error as it should be temporary and " +
+				"proceed with the migration again.")
+		case err == nil && revErr == nil:
+			log.Info("Migration of ENI datapath successful")
```
nit: If I follow right, this will print regardless of whether migration was actually performed. Would be nice to only print if action is taken.

christarazi replied:
Thanks. Here's the diff to fix this:

```diff --git a/daemon/daemon_main.go b/daemon/daemon_main.go
index 21aeb9133..5c72345b7 100644
--- a/daemon/daemon_main.go
+++ b/daemon/daemon_main.go
@@ -1337,7 +1337,7 @@ func runDaemon() {
 	// reasons as the API being served. We want to ensure that this migration
 	// logic runs before any endpoint creates.
 	if option.Config.IPAM == option.IPAMENI {
-		err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
+		migrated, err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
 		switch {
 		case err != nil && revErr != nil:
 			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. " +
@@ -1346,8 +1346,8 @@ func runDaemon() {
 			log.WithError(err).Warning("Failed to migrate ENI datapath. The original datapath has been reinstated. " +
 				"Connectivity has not been affected. Consider investigating the error as it should be temporary and " +
 				"proceed with the migration again.")
-		case err == nil && revErr == nil:
-			log.Info("Migration of ENI datapath successful")
+		case migrated > 0 && err == nil && revErr == nil:
+			log.Infof("Migration of ENI datapath successful, %d rules were migrated", migrated)
 		}
 	}
 
diff --git a/pkg/datapath/linux/routing/migrate.go b/pkg/datapath/linux/routing/migrate.go
index a179ffb0c..3a1936a26 100644
--- a/pkg/datapath/linux/routing/migrate.go
+++ b/pkg/datapath/linux/routing/migrate.go
@@ -24,14 +24,16 @@ import (
 )
 
 // MigrateENIDatapath migrates the egress rules inside the Linux routing policy
-// database (rpdb) for ENI IPAM mode. Because this migration must be
-// all-or-nothing, otherwise breaking connectivity, this function returns two
-// errors: one for the overall migration and one while reverting the successful
-// operation during the migration. For example, if there are 3 egress rules to
-// delete and 2 successfully get deleted, but the last fails, then what will be
-// returned is the error that occurred while deleting the last rule, and any
-// error that occurred while reverting the 2 rules that were successfully
-// deleted (meaning adding those 2 rules).
+// database (rpdb) for ENI IPAM mode. It will return the number of rules that
+// were successfully migrated. Because this migration must be all-or-nothing,
+// otherwise breaking connectivity, this function also returns two errors: one
+// for the overall migration and one while reverting the successful operation
+// during the migration. For example, if there are 3 egress rules to delete and
+// 2 successfully get deleted, but the last fails, then what will be returned
+// is the error that occurred while deleting the last rule, and any error that
+// occurred while reverting the 2 rules that were successfully deleted (meaning
+// adding those 2 rules). In an error case, zero will always be returned as the
+// number of rules actually migrated would be zero.
 //
 // We rely on the ability to fetch the CiliumNode resource because we need to
 // fetch the number associated with the ENI device. The CiliumNode resource
@@ -47,11 +49,10 @@ import (
 // ENI has an ifindex that equals the main routing table number (253-255),
 // causing the rules and routes to be created using the wrong table ID, which
 // could end up blackholing most traffic on the node.
-func MigrateENIDatapath(getter interfaceDB) (error, error) {
+func MigrateENIDatapath(getter interfaceDB) (int, error, error) {
 	rules, err := netlink.RuleList(netlink.FAMILY_V4)
 	if err != nil {
-		return fmt.Errorf("failed to list rules, cannot migrate ENI routing table: %w", err),
-			nil
+		return 0, fmt.Errorf("failed to list rules, cannot migrate ENI routing table: %w", err), nil
 	}
 
 	// This operation will be done over two phases. If either fails, we will
@@ -64,6 +65,9 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 		// migrating.
 		revertStack revert.RevertStack
 
+		// Number of rules successfully migrated. On errors, this will be zero.
+		migrated int
+
 		// Error while migrating and error while reverting, respectively.
 		overallErr, revertErr error
 	)
@@ -130,7 +134,7 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
 		}
 
-		return overallErr, revertErr
+		return 0, overallErr, revertErr
 	}
 
 	for _, r := range rules {
@@ -180,6 +184,10 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 				break
 			}
 
+			// We count the successful cases here because it means we have
+			// passed both phases of the migration.
+			migrated++
+
 		case linux_defaults.RulePriorityEgressv2:
 			// Downgrade case. Unsupported for now.
 		}
@@ -191,10 +199,10 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
 		}
 
-		return overallErr, revertErr
+		return 0, overallErr, revertErr
 	}
 
-	return nil, nil
+	return migrated, nil, nil
 }
 
 // retrieveTableIDFromIfIndex computes the correct table ID based on the
diff --git a/pkg/datapath/linux/routing/migrate_test.go b/pkg/datapath/linux/routing/migrate_test.go
index 5f6cc2f2a..7a36176e1 100644
--- a/pkg/datapath/linux/routing/migrate_test.go
+++ b/pkg/datapath/linux/routing/migrate_test.go
@@ -116,9 +116,10 @@ func (m *MigrateSuite) assertRoutingTable(c *C) {
 	}
 
 	// (2) Make the call to modifying the routing table.
-	err, revErr := MigrateENIDatapath(m)
+	migrated, err, revErr := MigrateENIDatapath(m)
 	c.Assert(err, IsNil)
 	c.Assert(revErr, IsNil)
+	c.Assert(migrated, Equals, 1)
 
 	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
 		Table: index,
```

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,404 @@
+// We rely on the ability to fetch the CiliumNode resource because we need to
+// fetch the number associated with the ENI device. The CiliumNode resource
+// contains this information in the Status field. This fetch is abstracted away
+// in the interfaceDB getter to avoid brining in K8s logic to this low-level
```
```suggestion
// in the interfaceDB getter to avoid bringing in K8s logic to this low-level
```

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,404 @@
+			}
+
+			revertStack.Push(func() error {
+				if err := netlink.RouteDel(&route); err != nil {
```
I suspect we're in "good enough" territory here, but note that "RouteReplace" may overwrite an existing route, which means that the corresponding revert operation would be to reinstate the replaced route, not to delete the new one.

That said, I think we're probably going a bit too far off the beaten track considering that we assume Cilium is the one administering these route tables?

joestringer replied:
(Unless we're specifically worried about the case where the ifindex and the new route table are the same number..?)

joestringer reviewed (approved):
LGTM. Remaining nits are I think sufficient to address in the master version along with the missing bits for better netlink library use, Azure, etc.

christarazi committed:
docs: Add upgrade docs for egress-multi-home-ip-rule-compat

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
api: Expose egress-multi-home-ip-rule-compat flag

This is important for use in the CNI to decide whether to use the new
ENI datapath (see previous commit for context) or the original datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -316,6 +316,24 @@ Annotations:
+
 .. _current_release_required_changes:
 
 .. _1.7_required_changes:
```
nit: Should probably make these link targets above the new section so that users don't accidentally skip over these notes when upgrading.

We can also call this out for ENI users in the release notes.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -316,6 +316,24 @@ Annotations:
+It defaults to ``false`` meaning it will instruct Cilium to **not** maintain
+compatibility and perform a migration to the new ENI datapath, upon Cilium
+upgrade. This operation will modify the routing table to fix a potential
+collision.
```
```suggestion
upgrade. This operation will modify the routing table to fix a collision
that in rare cases can cause connectivity loss to the node.
```

joestringer replied:
As discussed, would be good to have a step here to verify success of the migration.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -316,6 +316,24 @@ Annotations:
+Cilium so that it can rollback the migration. This ensures that the new ENI
+datapath is rolled back to a state where it is compatible with the older
+version of Cilium.
+
```
```suggestion
If the migration was already completed and you wish to downgrade to
v1.7.12 or earlier, you must first set the flag to ``true`` and then restart
Cilium so that it can rollback the migration. This ensures that the new ENI
datapath is rolled back to a state where it is compatible with the older
version of Cilium. Once this has migrated back to the compatibility mode,
further downgrade to v1.7.12 or earlier can be performed without connectivity
disruption to active pods.

```

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -316,6 +316,24 @@ Annotations:
+1.7.13 Upgrade Notes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A new flag ``--egress-multi-home-ip-rule-compat`` was introduced to allow
```
I wonder if it'd be better to lead with the fix, then describe the flag? EG:

> In rare cases, Cilium 1.7.12 or earlier when configured with ENI IPAM mode
could disrupt connectivity for the node when new ENI devices are created on
the node. This version fixes the issue and introduces a new flag to handle compatibility.

joestringer reviewed (commented):
We discussed the "revert" case out of band, and I have posted some notes on the docs portion here. Otherwise I took a look at all of the new implementation and it LGTM. I appreciate in particular some of the negative testing to validate the rollback scenarios.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -320,6 +320,37 @@ Annotations:
+The flag defaults to ``false`` meaning it will instruct Cilium to **not**
+maintain compatibility and perform a migration to the new ENI datapath, upon
+Cilium upgrade. This operation will modify the routing table to fix a collision
+that in rare cases can cause connectivity loss to the node.
```
On second read, `not maintain compatibility` sounds scary if you don't know what's going on. It sounds like there will be network impact. If everything goes well here, there should be no user interaction necessary. I wonder if we can clarify this a bit more.

```suggestion
The flag defaults to ``false`` meaning it will instruct Cilium to perform a
migration to the new ENI datapath upon
Cilium upgrade. This operation will modify the routing table to fix a collision
that in rare cases can cause connectivity loss to the node.
```

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -320,6 +320,37 @@ Annotations:
+pods.
+
+When upgrading to this version, you can verify the success of the migration by
+inspecting for a message like below:
```
```suggestion
When upgrading to this version, you can verify the success of the migration by
inspecting Cilium agent logs for a message like below:
```

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1335,37 @@ func runDaemon() {
+		switch {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. " +
+				"Connectivity is not guaranteed and manual intervention is required.")
```
We had some discussion yesterday about the `Fatal()` here, the concern being that at a high level if somehow migration fails and the rollback also fails, then the node will be left in a "hybrid" state with both new and old route rules/tables and the upgrade will fail with the Cilium pod exiting out fatally due to this line.

Then when downgrading, if we also attempt to migrate back to the old scheme and that fails, in the previous version that would also try to rollback whatever changes were made while trying to migrate back to the old scheme, and then that deployment would also trigger the `Fatal()` here.

So in the worst case, the upgrade would fail with crashing Cilium, then the instructions tell the user to disable the compat flag and maybe that succeeds but worst case it also fails and Cilium crashes and then the user is probably left guessing exactly what to do but presumably would roll back to v1.7.12 or earlier.

I didn't yet review the most recent changes since last night, just wanted to check whether this particular concern was considered in the last changes.

christarazi replied:
Just to sync here from our offline discussion. 

We discussed the need for `Fatal` and determined that it is not necessary to `Fatal`. This is because we can guarantee an atomic datapath migration. Due to the order in which we migrate to another datapath, an endpoint can either have the old datapath or the new datapath, at any point, regardless of failures. (It is possible to have both datapaths in-place as [described here](https://github.com/cilium/cilium/pull/14337#discussion_r546999794), but that doesn't cause any issues because the rule and routes only differ by table ID, but have the same routing info anyway).

Instead, we will warn the user using log msgs to indicate if anything had gone wrong and when a warning is serious enough to warrant trying the migration again. The worse case is when the user doesn't heed the warnings, leading to potential future conflicts with new endpoints. The warnings advise cleaning up certain state that was left over in case of a failure in cleaning up in the migration.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1335,42 @@ func runDaemon() {
+					"and %d failed to migrate completely. "+
+					"The original datapath is still in-place, however it is recommended to retry the migration.",
+				migrated, failed, partial)
+			log.Error(msg)
```
As a general note, I think ideally we should be using structured logging for any of these variable values. This helps users with structured logging databases to find the same log across multiple instances, and helps us when grepping around the codebase for the message as we don't have to do any guesswork about which part is a hardcoded string vs. variable data.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1328,6 +1335,42 @@ func runDaemon() {
+					"Found %d partially migrated endpoints, which indicate a previously attempted failed migration. "+
+						"Consider setting --%s=%v and restarting Cilium to rollback or manually intervening.",
+					partial, option.EgressMultiHomeIPRuleCompat, !option.Config.EgressMultiHomeIPRuleCompat)
+			}
```
I'm struggling a little bit to wrap my head around why "partial" migration is any different.

On an endpoint-by-endpoint basis, we either configure the new mode and successfully delete the old rule (completing migration), or some point up until the delete (including the delete) fails, in which case that endpoint is not migrated.

From a user action perspective, I don't think there's a difference between partial and unmigrated instances, either way they need to either retry migration by restarting the daemon or figure out how to manually clean up old routes.

Why do we need dedicated logic to handle this case? Why not just retry the migration from whatever state is present?

Or, put another way, the route rule with the old priority is the switch as to whether we should perform migration for that endpoint. If it's present, we should attempt migration. If not, there is no migration to do. In the latter case, maybe in some rare condition there are still also corresponding route tables lying around, but as long as there are no remaining references to the route table we can just try to clean them up, or worst case log a warning about them. Presence of those extra route tables should not impact the choice to migrate.

joestringer replied:
I think the TL;DR of my commentary here is "Can we just delete the code that handles partial migration?"

christarazi replied:
They're just random state leftover, so yes we can just remove that code. I think we previously wanted to handle them because we didn't quite understand / validate that regardless of failures in the migration, there will always be the new or original datapath in-place. Given that, this code becomes of little value and it doesn't matter to the end user that there are partially migrated rules; the endpoints' connectivity is still working anyway and at some point, we've already logged a msg about the underlying issue that caused them to not be deleted.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	// in-place to direct traffic for the endpoint. There is a slight nuance
+	// here; the older rule has a _higher_ priority which means it would be
+	// selected by Linux first, before reaching the new rule. Therefore, if we
+	// fail at (3), keep in mind we do have the new state in-place, but it goes
```
This nuance is actually quite interesting for downgrade vs. upgrade.

In the upgrade case, the datapath changeover occurs in (3). In the downgrade case, because the route is a higher priority, the datapath changeover actually occurs in (2).

I haven't thought through whether this has any impact on the logic below yet.

christarazi replied:
**Warning**: it will get slightly confusing, but the TLDR is that the nuance doesn't really matter in the end. Connectivity to the endpoint will continue un-disrupted.

I updated the comment to clarify. In both cases, it is possible to have both datapaths in-place and the rule priority will be the discriminator. 

To expand, if we fail at (3), then:

- Upgrade:
  - the original datapath is still in-place, using the old rule with a _higher_ priority. So no issues.
- Downgrade:
  - the downgrade was essentially completed and we'll be using the downgraded version because the "newer" rule has a _higher_ priority.

You could consider the downgrade case "more successful" if you go by which datapath is being used. In the upgrade case, the new datapath is not being used, but rather the original because it has a rule with a _higher_ priority.

joestringer replied:
Isn't that a subtle difference though? On upgrade, failure at (3) means the datapath has not been upgraded, so you *should* clean up all the state we put in place during upgrade, since it's not in use. On downgrade, failure at (3) means that the datapath has in fact been downgraded successfully to the older version, so then you *should not* remove the newly instantiated (downgraded) version of the datapath, otherwise you'll upgrade it again.

christarazi replied:
Offline sync:

What Joe is pointing out is that we should have different logic for handling (3). Specifically for downgrade, we don't revert the work we've done up to this point; we can just call it a success, because the endpoint is using the downgraded datapath, regardless of the leftover state.

In contrast, for upgrades, we should attempt to revert because the leftover state could block the user's next attempt at retying the upgrade.

This will be implemented.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	var migrated, failed int
+
+	if isUpgrade {
+		for _, r := range v1Rules {
```
Now that there's less state sharing between iterations of this loop, is it possible to factor the inner logic into a separate `m.upgradeRule()` function so that we can reason about this current function from a high level of interactions between iterations of the loop, then reason about the single-endpoint case in a dedicated function?

joestringer replied:
Note that if a dedicated function like this returned an error, then the `failed++` / `continue` statements would not need to be duplicated either, the top level loop would look something like:

```
        if isUpgrade {
                for _, r := range v1Rules {
                        if err := m.upgradeRule(r); err != nil {
                                log.WithField("rule", r).Warning("...")
                                failed++
                                continue
                        }
                        migrated++
                }
        } else if isDowngrade {
...
```

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+				scopedLog.WithError(err).
+					Error("Failed to migrate endpoint to new ENI datapath due to failure in listing routes associated with rule. " +
+						"Previous datapath is still intact and endpoint connectivity is not affected.")
+
```
If there was a separate `m.upgradeRule()` function to handle the inner logic of the loop, many of these error conditions could simply return a `fmt.Errorf(...)` with the unique bits of these error messages, and we would need only a few lines that actually do the logging. I think this would be easier to read since so many of the lines of this function are similar-looking error messages.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+					msg := "Failed to migrate endpoint to new ENI datapath and failed to cleanup new rule and routes. " +
+						"Previous datapath is still intact and will continue to operate for the endpoint. " +
+						"It is advised to investigate these errors."
+					scopedLog.WithError(err).WithField("revertError", revErr).Error(msg)
```
These revert error cases I think should still be a dedicated log line and should be `Warning(...)` and should *not* cause an error to be triggered by themselves.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+				}
+
+				failed++
+				continue
```
With a dedicated function for the inner logic, these lines would boil down to:

```
return fmt.Errorf("Failed to insert new routing rule")
```

Which is clear, concise, and easier to follow than trying to scan through and skip past the multiline logging logic handling.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	return link.Attrs().Index, nil
+}
+
+func (m *migrator) copyRoutes(routes []netlink.Route, from, to int) (revert.RevertStack, error) {
```
There's a particular assumption that this function makes which is not called out, and is subtly different from some other code:

This function could return three possible results:
* No reverstack and an error
* Revertstack (to revert first successful operation) and an error
* Reverstack (to revert all operations) and no error

Therefore, the caller *must* run `revertstack.Revert()` while handling the error and if they miss that, it's a bug. I note that the code today seems to do this pretty consistently so I think it's correct. That said, this is an assumption the code makes.

The caller can be simplified slightly by streamlining the possible return cases to:
* No reverstack and an error
* Reverstack (to revert all operations) and no error

There's multiple ways to handle this, one would be to add a defer which checks the error result & automatically runs the revertstack if an error occurs. Another common way in other languages is using a goto for a dedicated error handling section at the end of the function.

I don't think we necessarily have to switch to the latter approach, but ideally we should call out this assumption above the function signature so that the expectations for the caller are more clear. (Conversely, if we *do* change this, then all callers must be checked for their usage of the reverstack to conform to whichever model this function assumes)

christarazi replied:
Updated the godoc on `copyRoutes` and `createNewRule`.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+				} else {
+					scopedLog.WithError(err).
+						Info("Failed to downgrade endpoint to original ENI datapath due to failure deleting old rule, " +
+							"but was successfully reverted. Previous datapath is still intact and will continue to " +
```
This is not true for the `deleteOldRule()` case for the downgrade. Given that inserting the "new" (in this case, v1 scheme) rule involves creating a higher priority rule, the 'previous' datapath is already out of the picture.

(This is the upgrade vs. downgrade distinction between migration occurring in (2) vs. (3) [here](https://github.com/cilium/cilium/pull/14337#discussion_r546106322))

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	})
+
+	return revertStack, nil
+}
```
I don't think we need the revert logic here any more since it'd be ignored in the caller?

christarazi committed:
routing: Refactor helper to run function in netns

This makes it usable for an upcoming commit which adds a new test suite
to this packag.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
routing: Remove unnecessary debug logs from test

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,623 @@
+// returns a RevertStack and an error. The RevertStack contains functions that
+// would revert all the successful operations that occurred in this function.
+// The caller of this function should be aware to ensure the stack is reverted
+// upon an error.
```
Let's not make this ambiguous ;-)

```suggestion
// The caller of this function MUST revert the stack when this function returns an error.
```

Same for `createNewRule()`.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,623 @@
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(downgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to delete old rule: %w", err)
```
I think that this should actually 
```suggestion
		return nil, nil
```

Since the rule insertion above actually made the downgrade effective. Thoughts?

joestringer replied:
^^^ Something I missed in my original post here, if so we should also remove the RevertStack handling logic a few lines above.

christarazi replied:
Yes, that's correct. Will fix.

christarazi replied:
On second thought, would it be useful for the user to be informed of this, so that they can cleanup the leftover state themselves?

joestringer replied:
Yes, I think we should log like we do for the other "failed cleanup" steps.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,623 @@
+	}
+
+	return nil
+}
```
nit: This function doesn't seem to add anything any more, presumably the caller could just directly call `m.rpdb.RuleDel(...)`?

joestringer replied:
(but if you think it's easier to read with this function I'm also fine to just ignore this feedback)

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -320,6 +320,37 @@ Annotations:
+The flag defaults to ``false`` meaning it will instruct Cilium to perform a
+migration to the new ENI datapath upon Cilium upgrade. This operation will
+modify the routing table to fix a collision that in rare cases can cause
+connectivity loss to the node.
```
I'm not sure the second sentence adds anything now that the first paragraph explains the bug and what the flag is?

```suggestion
The flag defaults to ``false`` meaning it will instruct Cilium to perform a
migration to the new ENI datapath upon Cilium upgrade. 
```

christarazi replied:
Good catch, I think this was a leftover from previous feedback and the msg kinda got duplicated.

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,623 @@
+// database (RPDB) for ENI IPAM mode. It will return the number of rules that
+// were successfully migrated and the number of rules we've failed to migrated.
+// A -1 is returned for the failed number of rules indicating we couldn't even
+// start the migration.
```
nit, without conditionals it sounds like `failed` is always `-1`:

```suggestion
// A -1 may be returned for the failed number of rules indicating we couldn't even
// start the migration.
```

or

```suggestion
// A -1 is returned for the failed number of rules if we couldn't even
// start the migration.
```

christarazi replied:
Taking the latter, thanks

joestringer commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,623 @@
+			// The priority has been updated to 111 and the table ID is 12
+			// because the interface number is 2 plus the routing table offset
+			// (linux_defaults.RouteTableInterfacesOffset). See copyRoutes()
+			// for what happens with routes.
```
nit, these comments are probably more useful in the `upgradeRule()` logic (and corresponding `downgradeRule()` logic) now that there's a dedicated function. But they're super useful to compare against to reason about the logic :+1:

christarazi replied:
Good point, fixed.

joestringer reviewed (approved):
:100: Just a few minor nits but this is looking good. Main function feedback is just on adjusting the downgrade path so we call it successful after the higher priority rule is put in place, regardless of any subsequent failure in operations.

twpayne commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -54,6 +54,7 @@ cilium-agent [flags]
       --disable-k8s-services                          Disable east-west K8s load balancing by cilium
       --dns-max-ips-per-restored-rule int             Maximum number of IPs to maintain for each restored DNS rule (default 1000)
       --egress-masquerade-interfaces string           Limit egress masquerading to interface selector
+      --egress-multi-home-ip-rule-compat              Use a new scheme to store rules and routes under ENI and Azure IPAM modes, if false. Otherwise, it will use the old scheme.
```
nit: The words "new" and "old" will likely only make sense to people who have read this PR. Is it possible to find a more descriptive description. Maybe something like "Offset routing table IDs under ENI and Azure IPAM modes to avoid collisions with reserved table IDs."

twpayne commented on daemon/daemon_main.go:
```diff
@@ -770,6 +773,10 @@ func init() {
 	option.BindEnv(option.EnableAPIRateLimit)
 
+	flags.Bool(option.EgressMultiHomeIPRuleCompat, false,
+		"Use a new scheme to store rules and routes under ENI and Azure IPAM modes, if false. Otherwise, it will use the old scheme.")
```
nit: see nit about flag description in generated `Documentation/cmdref/cilium-agent.md` above.

twpayne commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -23,6 +23,12 @@ const (
+	// RouteTableInterfacesOffset is the offset for the per-ENI routing tables.
+	// Each ENI interface will have its own table starting with this offset. It
+	// is 10 because it is highly unlikely to collide with the main routing
+	// table which is between 253-255. See ip-route(8).
```
nit: table id 0 is also reserved (as well as 253-255).

twpayne commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	//
+	// If we fail at (4), then the old rule will have been deleted and the new
+	// state is in-place, which would be servicing the endpoint. The old routes
+	// would just be leftover state to be cleaned up at a later point.
```
Is there any mechanism currently in place to clean up rules and routes left over from a partial migration "at a later point"? It may not be needed in practice of course.

christarazi replied:
There is not. There are certain scenarios that are more risky than others if we fail to cleanup and we warn the user of those specific cases. Otherwise, we perform a best-effort to cleanup. If something fails, we'll tell the user that they should clean it up, but otherwise datapath connectivity is unaffected.

twpayne commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+			scopedLog.WithError(err).WithField("routes", routes).
+				Warnf("Failed to cleanup after successfully migrating endpoint to %s ENI datapath. "+
+					"It is recommended that theses routes are cleaned up, as it is possible in the future "+
+					"to collide with another endpoint with the same IP.", version)
```
How would the user actually perform the cleanup? By running `ip route del` manually to remove routes with priority 110 for example? It's maybe worth giving the user a hint here.

twpayne commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	downgradeRevertWarning = "Reverting the new ENI datapath failed. However, both the new and previous datapaths are still intact. " +
+		"Endpoint connectivity should not be affected. It is advised to retry the migration."
+	downgradeFailedRuleDeleteWarning = "Downgrading the datapath has succeeded, but failed to cleanup the original datapath. " +
+		"It is advised to manually remove the old rule."
```
Maybe note that old rules have priority 110.

twpayne commented on pkg/datapath/linux/routing/migrate.go:
```diff
@@ -0,0 +1,628 @@
+	// pkg/aws/eni/limits.go). Because the interface number is monotonically
+	// increasing and the lowest available number is reused when devices are
+	// added / removed. This means that the max possible table ID is 25.
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
```
Is it worth adding a check here that the returned table ID does not actually conflict with the main routing table? Of course, right now the maximum number of ENI devices on a node is 15, but this may change in the future. Adding a check here would catch the problem early rather than inadvertently colliding with the main routing table when a machine in the future gets its 243rd ENI.

christarazi replied:
Added a check which will warn the user in this case.

twpayne commented on pkg/datapath/linux/routing/migrate_test.go:
```diff
@@ -0,0 +1,575 @@
+// Copyright 2016-2020 Authors of Cilium
```
nit: Copyright year needs to be updated to 2021 I think. Same with other files touched in this PR.

twpayne commented on pkg/datapath/linux/routing/migrate_test.go:
```diff
@@ -0,0 +1,575 @@
+	return filterRulesByPriority(rules, prio), nil
+}
+
+func (m *MigrateSuite) defaultNetlinkMock() {
```
Not for this PR, but there are a number of projects that generate Go mocks, avoiding having to manually create them. One example is [gomock](https://github.com/golang/mock).

twpayne reviewed (approved):
LGTM. A few nits mainly around UX, but they're minor so feel free to ignore them.
